File: frontend/src/characters/characterSheetV2SpellProgression.ts
import { characterCreationRules } from '../rules/generated/characterCreationRules';
import { levelUpRules } from '../rules/generated/levelUpRules';
import type {
  CharacterSheetV2Spell,
  CharacterSpellcastingInput,
  SpellSelectionInput,
} from './characterSheetV2';

export type SpellReconstructionInput = {
  classIndex: string;
  subclassIndex: string | null;
  level: number;
  abilityModifier: number;
  input: CharacterSpellcastingInput;
  activeFeatureIds?: string[];
  raceGrantedCantripIndexes?: string[];
  classGrantedCantripIndexes?: string[];
};

export type SpellReconstructionResult = {
  spells: CharacterSheetV2Spell[];
  preparedSpellIds: string[];
  alwaysPreparedSpellIds: string[];
};

type CanonicalSpellcasting = {
  mode: Exclude<CharacterSpellcastingInput['mode'], 'none'>;
  ability: 'intelligence' | 'wisdom' | 'charisma';
  cantripsKnown?: number;
  spellsKnown?: number;
  preparedFormula?: string | null;
  replacementLimit: number;
  initialSpellbookSpells: number;
  wizardSpellbookAdditions?: number;
  availableSpellLevels: readonly number[];
};

export const reconstructSpellcastingV2 = (context: SpellReconstructionInput): SpellReconstructionResult => {
  const classRule = levelUpRules.classes.find(({ index }) => index === context.classIndex);
  const target = classRule?.levels.find(({ level }) => level === context.level);
  if (!classRule || !target) throw new Error('canonical Class level is unavailable');
  const expectedMode = target.spellcasting?.mode ?? 'none';
  if (context.input.mode !== expectedMode) throw new Error(`spellcasting mode must be ${expectedMode}`);
  const granted = [...raceGrantedCantrips(context), ...classGrantedCantrips(context)];
  if (expectedMode === 'none') return { spells: granted, preparedSpellIds: [], alwaysPreparedSpellIds: [] };
  const targetCasting = target.spellcasting as CanonicalSpellcasting | null;
  if (!targetCasting) throw new Error('spellcasting progression is unavailable');
  if (context.input.mode === 'none') throw new Error(`spellcasting mode must be ${expectedMode}`);
  const slotOverrides = context.input.slotOverride ?? [];
  if (slotOverrides.some(({ level }) => !(targetCasting.availableSpellLevels as readonly number[]).includes(level))) {
    throw new Error('spell slot override level is unavailable');
  }

  const cantrips = resolveDistinct(context.input.cantrips, context, context.level, true);
  if (cantrips.length !== (targetCasting.cantripsKnown ?? 0)) {
    throw new Error(`choose exactly ${targetCasting.cantripsKnown ?? 0} cantrips`);
  }

  let ordinary: CharacterSheetV2Spell[] = [];
  let preparedSpellIds: string[] = [];
  if (context.input.mode === 'known' || context.input.mode === 'pact-known') {
    ordinary = reconstructKnown(context, context.input.levels);
  } else if (context.input.mode === 'prepared') {
    ordinary = resolveDistinct(context.input.prepared, context, context.level, false).map((spell) => ({ ...spell, state: 'prepared' }));
    const wanted = preparedLimit(targetCasting.preparedFormula, context.abilityModifier, context.level);
    if (ordinary.length !== wanted) throw new Error(`choose exactly ${wanted} prepared spells`);
    preparedSpellIds = ordinary.map(({ id }) => id);
  } else {
    const initial = resolveDistinct(context.input.initialSpellbook, context, 1, false);
    const initialCount = classRule.levels[0].spellcasting?.initialSpellbookSpells ?? 0;
    if (initial.length !== initialCount) throw new Error(`choose exactly ${initialCount} initial spellbook spells`);
    if (initial.some(({ level }) => level !== 1)) throw new Error('initial Wizard spellbook spells must be level 1');
    const additions = context.input.additions;
    if (additions.length !== Math.max(0, context.level - 1) || additions.some((entry, index) => entry.level !== index + 2)) {
      throw new Error(`Wizard spellbook additions must cover every level 2 through ${context.level}`);
    }
    ordinary = [...initial];
    for (const addition of additions) {
      const levelRule = classRule.levels.find(({ level }) => level === addition.level)?.spellcasting as CanonicalSpellcasting | null | undefined;
      const spells = resolveDistinct(addition.spells, context, addition.level, false);
      if (!levelRule || spells.length !== levelRule.wizardSpellbookAdditions) throw new Error(`Wizard level ${addition.level} requires exactly two additions`);
      appendUnique(ordinary, spells);
    }
    ordinary = ordinary.map((spell) => ({ ...spell, state: 'spellbook' }));
    const known = new Set(ordinary.map(({ id }) => id));
    if (new Set(context.input.preparedSpellIds).size !== context.input.preparedSpellIds.length || context.input.preparedSpellIds.some((id) => !known.has(id))) {
      throw new Error('Wizard prepared spells must be a distinct spellbook subset');
    }
    const wanted = preparedLimit(targetCasting.preparedFormula, context.abilityModifier, context.level);
    if (context.input.preparedSpellIds.length !== wanted) throw new Error(`choose exactly ${wanted} prepared spells`);
    preparedSpellIds = [...context.input.preparedSpellIds];
  }

  const automatic = automaticSpells(context);
  const ordinaryIdentities = new Set([...cantrips, ...ordinary, ...granted].map(spellIdentity));
  if (ordinaryIdentities.size !== cantrips.length + ordinary.length + granted.length || automatic.some((spell) => ordinaryIdentities.has(spellIdentity(spell)))) throw new Error('automatic or granted spell duplicates a normal selection');
  return {
    spells: [...granted, ...cantrips.map((spell) => ({ ...spell, state: 'known' as const })), ...ordinary, ...automatic],
    preparedSpellIds,
    alwaysPreparedSpellIds: automatic.map(({ id }) => id),
  };
};

const raceGrantedCantrips = (context: SpellReconstructionInput): CharacterSheetV2Spell[] => (context.raceGrantedCantripIndexes ?? []).map((index) => {
  const spell = characterCreationRules.spells.find((entry) => entry.index === index);
  const choice = characterCreationRules.raceChoices.find((entry) => entry.id === 'high-elf-cantrip');
  if (!spell || spell.level !== 0 || !choice?.allowedOptionIndexes.includes(index as never)) throw new Error('Race-granted cantrip is unavailable');
  return { id: `race-high-elf-cantrip-${index}`, canonicalIndex: spell.index, name: spell.name, level: spell.level, school: spell.school, castingTime: spell.castingTime, range: spell.range, components: [...spell.components], materialComponent: spell.material ?? null, duration: spell.duration, concentration: spell.concentration, ritual: spell.ritual, description: spell.description, higherLevelText: spell.higherLevel ?? null, state: 'known' as const, provenance: { kind: 'calculated' as const, ruleId: 'high-elf-cantrip' } };
});

const classGrantedCantrips = (context: SpellReconstructionInput): CharacterSheetV2Spell[] => (context.classGrantedCantripIndexes ?? []).map((index) => {
  const spell = characterCreationRules.spells.find((entry) => entry.index === index);
  const choice = characterCreationRules.classChoices.find((entry) => entry.id === 'circle-of-the-land-bonus-cantrip' && entry.classIndex === context.classIndex && entry.requiredSubclassIndex === context.subclassIndex);
  if (!spell || spell.level !== 0 || !choice?.options.some((option) => option.index === index)) throw new Error('Class-granted cantrip is unavailable');
  return { id: `class-circle-of-the-land-cantrip-${index}`, canonicalIndex: spell.index, name: spell.name, level: spell.level, school: spell.school, castingTime: spell.castingTime, range: spell.range, components: [...spell.components], materialComponent: spell.material ?? null, duration: spell.duration, concentration: spell.concentration, ritual: spell.ritual, description: spell.description, higherLevelText: spell.higherLevel ?? null, state: 'known' as const, provenance: { kind: 'calculated' as const, ruleId: 'circle-of-the-land-bonus-cantrip' } };
});

const reconstructKnown = (context: SpellReconstructionInput, levels: Extract<CharacterSpellcastingInput, { mode: 'known' | 'pact-known' }>['levels']) => {
  const classRule = levelUpRules.classes.find(({ index }) => index === context.classIndex)!;
  const progression = classRule.levels.filter(({ level, spellcasting }) => level <= context.level && spellcasting !== null);
  if (levels.length !== progression.length || levels.some((entry, index) => entry.level !== progression[index].level)) {
    throw new Error('known spell decisions must cover every spellcasting level');
  }
  const result: CharacterSheetV2Spell[] = [];
  let previousKnown = 0;
  progression.forEach((levelRule, index) => {
    const decision = levels[index];
    const casting = levelRule.spellcasting as CanonicalSpellcasting;
    const wantedLearned = index === 0 ? casting.spellsKnown ?? 0 : Math.max(0, (casting.spellsKnown ?? 0) - previousKnown);
    if (decision.learned.length !== wantedLearned) throw new Error(`level ${decision.level} learned spell count is invalid`);
    if (decision.replacements.length > casting.replacementLimit || index === 0 && decision.replacements.length > 0) throw new Error(`level ${decision.level} replacement count is invalid`);
    for (const replacement of decision.replacements) {
      const position = result.findIndex(({ id }) => id === replacement.removeSpellId);
      if (position < 0) throw new Error(`level ${decision.level} replacement removes an unknown spell`);
      const [added] = resolveDistinct([replacement.add], context, decision.level, false);
      if (added.id === replacement.removeSpellId || result.some((spell) => spellIdentity(spell) === spellIdentity(added))) throw new Error('replacement must add a distinct new spell');
      result.splice(position, 1, { ...added, state: 'known' });
    }
    appendUnique(result, resolveDistinct(decision.learned, context, decision.level, false).map((spell) => ({ ...spell, state: 'known' })));
    previousKnown = casting.spellsKnown ?? 0;
  });
  return result;
};

const resolveDistinct = (inputs: SpellSelectionInput[], context: SpellReconstructionInput, acquisitionLevel: number, cantrip: boolean) => {
  const ids = inputs.map(({ id }) => id);
  if (new Set(ids).size !== ids.length) throw new Error('spell selection IDs must be distinct');
  return inputs.map((input) => resolveSpell(input, context, acquisitionLevel, cantrip));
};

const resolveSpell = (input: SpellSelectionInput, context: SpellReconstructionInput, acquisitionLevel: number, cantrip: boolean): CharacterSheetV2Spell => {
  if (input.source === 'manual') {
    if (!input.importReason.trim()) throw new Error('manual spell requires an import reason');
    if ((input.level === 0) !== cantrip || !legalSpellLevel(context.classIndex, acquisitionLevel, input.level)) throw new Error('manual spell level is unavailable');
    return { id: input.id, canonicalIndex: null, name: input.name, level: input.level, school: input.school, castingTime: input.castingTime, range: input.range, components: [...input.components], materialComponent: input.materialComponent ?? null, duration: input.duration, concentration: input.concentration, ritual: input.ritual, description: input.description, higherLevelText: input.higherLevelText ?? null, state: 'known', provenance: { kind: 'imported', note: input.importReason } };
  }
  const spell = characterCreationRules.spells.find(({ index }) => index === input.index);
  if (!spell || (spell.level === 0) !== cantrip || !legalSpellLevel(context.classIndex, acquisitionLevel, spell.level) || !eligible(spell, context, acquisitionLevel)) {
    throw new Error(`${input.index} is unavailable for ${context.classIndex} level ${acquisitionLevel}`);
  }
  return { id: input.id, canonicalIndex: spell.index, name: spell.name, level: spell.level, school: spell.school, castingTime: spell.castingTime, range: spell.range, components: [...spell.components], materialComponent: spell.material ?? null, duration: spell.duration, concentration: spell.concentration, ritual: spell.ritual, description: spell.description, higherLevelText: spell.higherLevel ?? null, state: 'known', provenance: { kind: 'calculated', ruleId: 'spell-canonical' } };
};

const eligible = (spell: (typeof characterCreationRules.spells)[number], context: SpellReconstructionInput, level: number) =>
  (spell.classIndexes as readonly string[]).includes(context.classIndex) || spell.subclassMemberships.some((membership) =>
    membership.classIndex === context.classIndex && membership.subclassIndex === context.subclassIndex && membership.classLevel <= level && membership.kind === 'expanded');

const legalSpellLevel = (classIndex: string, classLevel: number, spellLevel: number) => spellLevel === 0 || Boolean((levelUpRules.classes.find(({ index }) => index === classIndex)?.levels.find(({ level }) => level === classLevel)?.spellcasting?.availableSpellLevels as readonly number[] | undefined)?.includes(spellLevel));

const automaticSpells = (context: SpellReconstructionInput): CharacterSheetV2Spell[] => {
  const selectedClass = levelUpRules.classes.find(({ index }) => index === context.classIndex);
  const activeFeatures = new Set([
    ...(context.activeFeatureIds ?? []),
    ...(selectedClass?.levels ?? []).filter(({ level }) => level <= context.level).flatMap(({ features }) => features.map(({ index }) => index)),
    ...(selectedClass?.subclasses.find(({ index }) => index === context.subclassIndex)?.featuresByLevel ?? [])
      .filter(({ level }) => level <= context.level).flatMap(({ features }) => features.map(({ index }) => index)),
  ]);
  return characterCreationRules.spells.flatMap((spell) => {
    const automatic = spell.subclassMemberships.some((membership) => membership.classIndex === context.classIndex && membership.subclassIndex === context.subclassIndex && membership.classLevel <= context.level && membership.kind === 'always-prepared' && membership.requiredFeatureIndexes.every((id) => activeFeatures.has(id)));
    return automatic ? [{ ...canonicalSpell(spell, `automatic-${spell.index}`), state: 'always-prepared' as const }] : [];
  });
};

const preparedLimit = (formula: string | null | undefined, abilityModifier: number, level: number) => {
  if (formula === 'max(1,abilityModifier+classLevel)') return Math.max(1, abilityModifier + level);
  if (formula === 'max(1,abilityModifier+floor(classLevel/2))') return Math.max(1, abilityModifier + Math.floor(level / 2));
  throw new Error('prepared spell formula is unavailable');
};

const appendUnique = (target: CharacterSheetV2Spell[], additions: CharacterSheetV2Spell[]) => {
  const identities = new Set(target.map(spellIdentity));
  for (const spell of additions) {
    const identity = spellIdentity(spell);
    if (identities.has(identity)) throw new Error(`duplicate spell selection ${spell.id}`);
    identities.add(identity);
    target.push(spell);
  }
};

const spellIdentity = (spell: CharacterSheetV2Spell) => spell.canonicalIndex === null ? `manual:${spell.id}` : `srd:${spell.canonicalIndex}`;

const canonicalSpell = (spell: (typeof characterCreationRules.spells)[number], id: string): CharacterSheetV2Spell => ({
  id,
  canonicalIndex: spell.index,
  name: spell.name,
  level: spell.level,
  school: spell.school,
  castingTime: spell.castingTime,
  range: spell.range,
  components: [...spell.components],
  materialComponent: spell.material ?? null,
  duration: spell.duration,
  concentration: spell.concentration,
  ritual: spell.ritual,
  description: spell.description,
  higherLevelText: spell.higherLevel ?? null,
  state: 'known',
  provenance: { kind: 'calculated', ruleId: 'spell-canonical' },
});

File: backend/internal/characters/character_sheet_v2_spell_progression.go
package characters

import (
	"fmt"

	"github.com/Inkala/rpg-companion/backend/internal/rules"
)

type SpellReconstructionInput struct {
	ClassIndex                 string
	SubclassIndex              string
	Level                      int
	AbilityModifier            int
	Input                      CharacterSpellcastingInput
	ActiveFeatureIDs           []string
	RaceGrantedCantripIndexes  []string
	ClassGrantedCantripIndexes []string
}

type SpellReconstructionResult struct {
	Spells                 []CharacterSheetV2Spell
	PreparedSpellIDs       []string
	AlwaysPreparedSpellIDs []string
}

func reconstructV2Spellcasting(context SpellReconstructionInput) (SpellReconstructionResult, error) {
	data, err := rules.Load()
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	class, ok := findClass(data.Classes, context.ClassIndex)
	if !ok {
		return SpellReconstructionResult{}, fmt.Errorf("canonical Class is unavailable")
	}
	target, ok := findLevel(class, context.Level)
	if !ok {
		return SpellReconstructionResult{}, fmt.Errorf("canonical Class level is unavailable")
	}
	expectedMode := "none"
	if target.Spellcasting != nil {
		expectedMode = target.Spellcasting.Mode
	}
	if context.Input.Mode != expectedMode {
		return SpellReconstructionResult{}, fmt.Errorf("spellcasting mode must be %s", expectedMode)
	}
	raceGranted, err := raceGrantedV2Cantrips(context.RaceGrantedCantripIndexes)
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	classGranted, err := classGrantedV2Cantrips(context)
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	granted := append(raceGranted, classGranted...)
	if expectedMode == "none" {
		return SpellReconstructionResult{Spells: granted, PreparedSpellIDs: []string{}, AlwaysPreparedSpellIDs: []string{}}, nil
	}
	casting := target.Spellcasting
	for _, override := range context.Input.SlotOverride {
		if !v2ContainsInt(casting.AvailableSpellLevels, override.Level) {
			return SpellReconstructionResult{}, fmt.Errorf("spell slot override level is unavailable")
		}
	}
	cantrips, err := resolveDistinctSpells(context.Input.Cantrips, context, context.Level, true)
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	cantripCount := 0
	if casting.CantripsKnown != nil {
		cantripCount = *casting.CantripsKnown
	}
	if len(cantrips) != cantripCount {
		return SpellReconstructionResult{}, fmt.Errorf("choose exactly %d cantrips", cantripCount)
	}
	for index := range cantrips {
		cantrips[index].State = "known"
	}

	ordinary := []CharacterSheetV2Spell{}
	preparedIDs := []string{}
	switch context.Input.Mode {
	case "known", "pact-known":
		ordinary, err = reconstructKnownSpells(context, class)
	case "prepared":
		ordinary, err = resolveDistinctSpells(context.Input.Prepared, context, context.Level, false)
		if err == nil {
			for index := range ordinary {
				ordinary[index].State = "prepared"
				preparedIDs = append(preparedIDs, ordinary[index].ID)
			}
			wanted, formulaErr := preparedSpellLimit(casting.PreparedFormula, context.AbilityModifier, context.Level)
			if formulaErr != nil {
				err = formulaErr
			} else if len(ordinary) != wanted {
				err = fmt.Errorf("choose exactly %d prepared spells", wanted)
			}
		}
	case "spellbook-prepared":
		ordinary, preparedIDs, err = reconstructWizardSpells(context, class)
	default:
		err = fmt.Errorf("unsupported spellcasting mode")
	}
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	automatic, err := automaticV2Spells(context, class)
	if err != nil {
		return SpellReconstructionResult{}, err
	}
	seen := map[string]bool{}
	for _, spell := range append(append(append([]CharacterSheetV2Spell{}, granted...), cantrips...), ordinary...) {
		identity := v2SpellIdentity(spell)
		if seen[identity] {
			return SpellReconstructionResult{}, fmt.Errorf("duplicate spell selection %s", spell.ID)
		}
		seen[identity] = true
	}
	for _, spell := range automatic {
		identity := v2SpellIdentity(spell)
		if seen[identity] {
			return SpellReconstructionResult{}, fmt.Errorf("automatic spell duplicates a normal selection")
		}
		seen[identity] = true
	}
	spells := append(append(append(granted, cantrips...), ordinary...), automatic...)
	alwaysIDs := make([]string, 0, len(automatic))
	for _, spell := range automatic {
		alwaysIDs = append(alwaysIDs, spell.ID)
	}
	return SpellReconstructionResult{Spells: spells, PreparedSpellIDs: preparedIDs, AlwaysPreparedSpellIDs: alwaysIDs}, nil
}

func classGrantedV2Cantrips(context SpellReconstructionInput) ([]CharacterSheetV2Spell, error) {
	creation, _ := rules.LoadCharacterCreation()
	var choice *rules.CharacterCreationChoice
	for index := range creation.ClassChoices {
		candidate := &creation.ClassChoices[index]
		if candidate.ID == "circle-of-the-land-bonus-cantrip" && candidate.ClassIndex == context.ClassIndex && candidate.RequiredSubclassIndex == context.SubclassIndex {
			choice = candidate
		}
	}
	result := make([]CharacterSheetV2Spell, 0, len(context.ClassGrantedCantripIndexes))
	for _, index := range context.ClassGrantedCantripIndexes {
		spell, ok := creation.FindSpellDetail(index)
		allowed := false
		if choice != nil {
			for _, option := range choice.Options {
				allowed = allowed || option.Index == index
			}
		}
		if !ok || spell.Level != 0 || !allowed {
			return nil, fmt.Errorf("Class-granted cantrip is unavailable")
		}
		canonicalIndex := spell.Index
		result = append(result, CharacterSheetV2Spell{ID: "class-circle-of-the-land-cantrip-" + index, CanonicalIndex: &canonicalIndex, Name: spell.Name, Level: spell.Level, School: spell.School, CastingTime: spell.CastingTime, Range: spell.Range, Components: append([]string(nil), spell.Components...), MaterialComponent: spell.Material, Duration: spell.Duration, Concentration: spell.Concentration, Ritual: spell.Ritual, Description: spell.Description, HigherLevelText: spell.HigherLevel, State: "known", Provenance: ValueProvenance{Kind: "calculated", RuleID: "circle-of-the-land-bonus-cantrip"}})
	}
	return result, nil
}

func raceGrantedV2Cantrips(indexes []string) ([]CharacterSheetV2Spell, error) {
	creation, _ := rules.LoadCharacterCreation()
	var choice *rules.RaceChoiceRule
	for index := range creation.RaceChoices {
		if creation.RaceChoices[index].ID == "high-elf-cantrip" {
			choice = &creation.RaceChoices[index]
		}
	}
	result := make([]CharacterSheetV2Spell, 0, len(indexes))
	for _, index := range indexes {
		spell, ok := creation.FindSpellDetail(index)
		if !ok || spell.Level != 0 || choice == nil || !v2Contains(choice.AllowedOptionIndexes, index) {
			return nil, fmt.Errorf("Race-granted cantrip is unavailable")
		}
		canonicalIndex := spell.Index
		result = append(result, CharacterSheetV2Spell{ID: "race-high-elf-cantrip-" + index, CanonicalIndex: &canonicalIndex, Name: spell.Name, Level: spell.Level, School: spell.School, CastingTime: spell.CastingTime, Range: spell.Range, Components: append([]string(nil), spell.Components...), MaterialComponent: spell.Material, Duration: spell.Duration, Concentration: spell.Concentration, Ritual: spell.Ritual, Description: spell.Description, HigherLevelText: spell.HigherLevel, State: "known", Provenance: ValueProvenance{Kind: "calculated", RuleID: "high-elf-cantrip"}})
	}
	return result, nil
}

func reconstructKnownSpells(context SpellReconstructionInput, class rules.Class) ([]CharacterSheetV2Spell, error) {
	progression := []rules.Level{}
	for _, level := range class.Levels {
		if level.Level <= context.Level && level.Spellcasting != nil {
			progression = append(progression, level)
		}
	}
	if len(context.Input.Levels) != len(progression) {
		return nil, fmt.Errorf("known spell decisions must cover every spellcasting level")
	}
	result := []CharacterSheetV2Spell{}
	previousKnown := 0
	for index, level := range progression {
		decision := context.Input.Levels[index]
		if decision.Level != level.Level {
			return nil, fmt.Errorf("known spell decisions must be deterministically ordered")
		}
		known := 0
		if level.Spellcasting.SpellsKnown != nil {
			known = *level.Spellcasting.SpellsKnown
		}
		wanted := known - previousKnown
		if wanted < 0 {
			wanted = 0
		}
		if len(decision.Learned) != wanted || len(decision.Replacements) > level.Spellcasting.ReplacementLimit || index == 0 && len(decision.Replacements) > 0 {
			return nil, fmt.Errorf("level %d spell decision count is invalid", level.Level)
		}
		for _, replacement := range decision.Replacements {
			position := -1
			for candidateIndex, candidate := range result {
				if candidate.ID == replacement.RemoveSpellID {
					position = candidateIndex
				}
			}
			if position < 0 {
				return nil, fmt.Errorf("replacement removes an unknown spell")
			}
			added, resolveErr := resolveDistinctSpells([]SpellSelectionInput{replacement.Add}, context, level.Level, false)
			if resolveErr != nil {
				return nil, resolveErr
			}
			if added[0].ID == replacement.RemoveSpellID || containsResolvedSpellIdentity(result, added[0]) {
				return nil, fmt.Errorf("replacement must add a distinct new spell")
			}
			added[0].State = "known"
			result[position] = added[0]
		}
		learned, resolveErr := resolveDistinctSpells(decision.Learned, context, level.Level, false)
		if resolveErr != nil {
			return nil, resolveErr
		}
		for candidateIndex := range learned {
			learned[candidateIndex].State = "known"
			if containsResolvedSpellIdentity(result, learned[candidateIndex]) {
				return nil, fmt.Errorf("duplicate learned spell")
			}
			result = append(result, learned[candidateIndex])
		}
		previousKnown = known
	}
	return result, nil
}

func reconstructWizardSpells(context SpellReconstructionInput, class rules.Class) ([]CharacterSheetV2Spell, []string, error) {
	initial, err := resolveDistinctSpells(context.Input.InitialSpellbook, context, 1, false)
	if err != nil {
		return nil, nil, err
	}
	initialCount := 0
	if class.Levels[0].Spellcasting != nil {
		initialCount = class.Levels[0].Spellcasting.InitialSpellbookSpells
	}
	if len(initial) != initialCount {
		return nil, nil, fmt.Errorf("choose exactly %d initial spellbook spells", initialCount)
	}
	for _, spell := range initial {
		if spell.Level != 1 {
			return nil, nil, fmt.Errorf("initial Wizard spellbook spells must be level 1")
		}
	}
	if len(context.Input.Additions) != context.Level-1 {
		return nil, nil, fmt.Errorf("Wizard spellbook additions must cover every later level")
	}
	ordinary := append([]CharacterSheetV2Spell{}, initial...)
	for index, addition := range context.Input.Additions {
		if addition.Level != index+2 {
			return nil, nil, fmt.Errorf("Wizard spellbook additions must be deterministically ordered")
		}
		level, _ := findLevel(class, addition.Level)
		spells, resolveErr := resolveDistinctSpells(addition.Spells, context, addition.Level, false)
		if resolveErr != nil {
			return nil, nil, resolveErr
		}
		if level.Spellcasting == nil || len(spells) != level.Spellcasting.WizardSpellbookAdditions {
			return nil, nil, fmt.Errorf("Wizard level %d requires exactly two additions", addition.Level)
		}
		for _, spell := range spells {
			if containsResolvedSpellIdentity(ordinary, spell) {
				return nil, nil, fmt.Errorf("duplicate spellbook spell")
			}
			ordinary = append(ordinary, spell)
		}
	}
	for index := range ordinary {
		ordinary[index].State = "spellbook"
	}
	seenPrepared := map[string]bool{}
	for _, id := range context.Input.PreparedSpellIDs {
		if seenPrepared[id] || !containsResolvedSpell(ordinary, id) {
			return nil, nil, fmt.Errorf("Wizard prepared spells must be a distinct spellbook subset")
		}
		seenPrepared[id] = true
	}
	target, _ := findLevel(class, context.Level)
	wanted, err := preparedSpellLimit(target.Spellcasting.PreparedFormula, context.AbilityModifier, context.Level)
	if err != nil || len(context.Input.PreparedSpellIDs) != wanted {
		return nil, nil, fmt.Errorf("choose exactly %d prepared spells", wanted)
	}
	return ordinary, append([]string(nil), context.Input.PreparedSpellIDs...), nil
}

func resolveDistinctSpells(inputs []SpellSelectionInput, context SpellReconstructionInput, acquisitionLevel int, cantrip bool) ([]CharacterSheetV2Spell, error) {
	result := make([]CharacterSheetV2Spell, 0, len(inputs))
	seen := map[string]bool{}
	for _, input := range inputs {
		if seen[input.ID] {
			return nil, fmt.Errorf("spell selection IDs must be distinct")
		}
		seen[input.ID] = true
		spell, err := resolveV2Spell(input, context, acquisitionLevel, cantrip)
		if err != nil {
			return nil, err
		}
		result = append(result, spell)
	}
	return result, nil
}

func resolveV2Spell(input SpellSelectionInput, context SpellReconstructionInput, acquisitionLevel int, cantrip bool) (CharacterSheetV2Spell, error) {
	if input.Source == "manual" {
		if !boundedText(input.ImportReason, 1000) || (input.Level == 0) != cantrip || !legalV2SpellLevel(context.ClassIndex, acquisitionLevel, input.Level) {
			return CharacterSheetV2Spell{}, fmt.Errorf("manual spell is unavailable or lacks import reason")
		}
		entry := CharacterSheetV2Spell{ID: input.ID, Name: input.Name, Level: input.Level, School: input.School, CastingTime: input.CastingTime, Range: input.Range, Components: append([]string(nil), input.Components...), Duration: input.Duration, Concentration: input.Concentration, Ritual: input.Ritual, Description: input.Description, State: "known", Provenance: ValueProvenance{Kind: "imported", Note: input.ImportReason}}
		if input.MaterialComponent != "" {
			value := input.MaterialComponent
			entry.MaterialComponent = &value
		}
		if input.HigherLevelText != "" {
			value := input.HigherLevelText
			entry.HigherLevelText = &value
		}
		return entry, nil
	}
	creation, _ := rules.LoadCharacterCreation()
	spell, ok := creation.FindSpellDetail(input.Index)
	if !ok || (spell.Level == 0) != cantrip || !legalV2SpellLevel(context.ClassIndex, acquisitionLevel, spell.Level) || !eligibleV2Spell(spell, context, acquisitionLevel) {
		return CharacterSheetV2Spell{}, fmt.Errorf("%s is unavailable for %s level %d", input.Index, context.ClassIndex, acquisitionLevel)
	}
	index := spell.Index
	return CharacterSheetV2Spell{ID: input.ID, CanonicalIndex: &index, Name: spell.Name, Level: spell.Level, School: spell.School, CastingTime: spell.CastingTime, Range: spell.Range, Components: append([]string(nil), spell.Components...), MaterialComponent: spell.Material, Duration: spell.Duration, Concentration: spell.Concentration, Ritual: spell.Ritual, Description: spell.Description, HigherLevelText: spell.HigherLevel, State: "known", Provenance: ValueProvenance{Kind: "calculated", RuleID: "spell-canonical"}}, nil
}

func eligibleV2Spell(spell rules.SpellDetail, context SpellReconstructionInput, level int) bool {
	if v2Contains(spell.ClassIndexes, context.ClassIndex) {
		return true
	}
	for _, membership := range spell.SubclassMemberships {
		if membership.ClassIndex == context.ClassIndex && membership.SubclassIndex == context.SubclassIndex && membership.ClassLevel <= level && membership.Kind == "expanded" {
			return true
		}
	}
	return false
}

func automaticV2Spells(context SpellReconstructionInput, class rules.Class) ([]CharacterSheetV2Spell, error) {
	activeFeatures := map[string]bool{}
	for _, featureID := range context.ActiveFeatureIDs {
		activeFeatures[featureID] = true
	}
	for _, level := range class.Levels {
		if level.Level <= context.Level {
			for _, feature := range level.Features {
				activeFeatures[feature.Index] = true
			}
		}
	}
	for _, subclass := range class.Subclasses {
		if subclass.Index == context.SubclassIndex {
			for _, group := range subclass.FeaturesByLevel {
				if group.Level <= context.Level {
					for _, feature := range group.Features {
						activeFeatures[feature.Index] = true
					}
				}
			}
		}
	}
	creation, _ := rules.LoadCharacterCreation()
	result := []CharacterSheetV2Spell{}
	for _, spell := range creation.Spells {
		for _, membership := range spell.SubclassMemberships {
			if membership.ClassIndex != context.ClassIndex || membership.SubclassIndex != context.SubclassIndex || membership.ClassLevel > context.Level || membership.Kind != "always-prepared" || !allV2FeaturesActive(membership.RequiredFeatureIndexes, activeFeatures) {
				continue
			}
			entry := canonicalV2Spell(spell, "automatic-"+spell.Index)
			entry.State = "always-prepared"
			result = append(result, entry)
			break
		}
	}
	return result, nil
}

func legalV2SpellLevel(classIndex string, classLevel, spellLevel int) bool {
	if spellLevel == 0 {
		return true
	}
	data, _ := rules.Load()
	class, ok := findClass(data.Classes, classIndex)
	if !ok {
		return false
	}
	level, ok := findLevel(class, classLevel)
	return ok && level.Spellcasting != nil && v2ContainsInt(level.Spellcasting.AvailableSpellLevels, spellLevel)
}

func preparedSpellLimit(formula *string, abilityModifier, level int) (int, error) {
	if formula == nil {
		return 0, fmt.Errorf("prepared spell formula is unavailable")
	}
	value := abilityModifier + level
	if *formula == "max(1,abilityModifier+floor(classLevel/2))" {
		value = abilityModifier + level/2
	} else if *formula != "max(1,abilityModifier+classLevel)" {
		return 0, fmt.Errorf("prepared spell formula is unavailable")
	}
	if value < 1 {
		value = 1
	}
	return value, nil
}

func containsResolvedSpell(spells []CharacterSheetV2Spell, id string) bool {
	for _, spell := range spells {
		if spell.ID == id {
			return true
		}
	}
	return false
}

func containsResolvedSpellIdentity(spells []CharacterSheetV2Spell, candidate CharacterSheetV2Spell) bool {
	identity := v2SpellIdentity(candidate)
	for _, spell := range spells {
		if v2SpellIdentity(spell) == identity {
			return true
		}
	}
	return false
}

func v2SpellIdentity(spell CharacterSheetV2Spell) string {
	if spell.CanonicalIndex == nil {
		return "manual:" + spell.ID
	}
	return "srd:" + *spell.CanonicalIndex
}

func canonicalV2Spell(spell rules.SpellDetail, id string) CharacterSheetV2Spell {
	index := spell.Index
	return CharacterSheetV2Spell{ID: id, CanonicalIndex: &index, Name: spell.Name, Level: spell.Level, School: spell.School, CastingTime: spell.CastingTime, Range: spell.Range, Components: append([]string(nil), spell.Components...), MaterialComponent: spell.Material, Duration: spell.Duration, Concentration: spell.Concentration, Ritual: spell.Ritual, Description: spell.Description, HigherLevelText: spell.HigherLevel, State: "known", Provenance: ValueProvenance{Kind: "calculated", RuleID: "spell-canonical"}}
}

func allV2FeaturesActive(required []string, active map[string]bool) bool {
	for _, id := range required {
		if !active[id] {
			return false
		}
	}
	return true
}
